Add fetch local file

zekth opened the pull request:
This fixes https://github.com/denoland/deno/issues/2150

I have 2 questions:
Here : https://github.com/denoland/deno/compare/master...zekth:fetch_local?expand=1#diff-cee0e68f0ac7840274b5171cf4df9dbeR410 i need to handle the `content-type` but it's in deno_std atm. Any idea? Also, how could i get the content-lenght?

ry commented on js/fetch.ts:
```diff
@@ -399,6 +401,28 @@ export async function fetch(
+      const f = openSync(m[1]);
+      const headers: [string, string][] = [];
+      // headers.push(["content-type", contentType(extname(m[1])) || "text/plain"])
+      headers.push(["content-type", "text/plain"]);
```
I think for now you don't need to return a content-type. It would be ideal if we could import https://github.com/denoland/deno_std/tree/master/media_types but that's not possible at the moment. Could you add a TODO comment that mentions this?

zekth replied:
Ok i was into the same approach. Sure i'll add it.

phil294 commented:
Cool. This repo is like a wish fairy.

I briefly looked into the [wikipedia article about file uri scheme](https://en.wikipedia.org/wiki/File_URI_scheme) and it said that syntax like `file://localhost/./myfile.txt` is valid too etc.

And what about other protocols like ftp?

Just bringing in ideas.

zekth commented:
> Cool. This repo is like a wish fairy.
> 
> I briefly looked into the [wikipedia article about file uri scheme](https://en.wikipedia.org/wiki/File_URI_scheme) and it said that syntax like `file://localhost/./myfile.txt` is valid too etc.
> 
> And what about other protocols like ftp?
> 
> Just bringing in ideas.

FTP may be added in another PR. Concerning all the schemes i'll dig into.

harrysarson commented:
Hi, I have a quick question:

As fetch is a asynchronous function, why is a synchronous open file function being used?

zekth commented:
> Hi, I have a quick question:
> 
> As fetch is a asynchronous function, why is a synchronous open file function being used?

Fair point. In fact the open is Synchronous but the reading is still async. Having the open as Sync is like building the header of the http request synchronously (as we do now). It would be interesting to measure the impact of both. BTW, i'll move it to async.

zekth commented:
@phil294 about HOST part of the URI it's not used at the moment because we are considering fetching a local file and not getting it throught the network. It can be a possibility but i think the implementation would change a bit. Also i've added a link to the RFC. Maybe a special section of the documentation may be added. For example the RFC is not very clear about the handling of relative path but a lot of interpreters are allowing it. ATM i think it fits the need.

See RFC8089 : https://tools.ietf.org/html/rfc8089

j-f1 commented on js/fetch.ts:
```diff
@@ -399,6 +401,28 @@ export async function fetch(
+      ) {
+        return new Response(404, [], 0);
+      } else {
+        return new Response(500, [], 0);
```
Is there any metadata associated with this error? If so, would it make sense to JSON-encode it and put it as the body of this response?

zekth replied:
As far as i know the errorkind notfound is just an enum for the error code. The only error i think would be not found or permission error. But just put a standard json encoded object like `{message:"file not found"}` would be interesting

harrysarson replied:
Would calling `JSON.stringify(e)` be a possibility (with try catch)?

zekth commented:
I've rebased the PR and added missing types for eslint.
Also for the return of the fetch in case of issue i've added this:
https://github.com/denoland/deno/pull/2152/files#diff-cee0e68f0ac7840274b5171cf4df9dbeR425

@kevinkassimo suggested to use `/dev/null` for unix and `NUL` for windows. Is there a way i open those filedescriptors and get the rid of it? any clue about this to land this PR?

zekth commented:
If you're satisfied with the rid:-1 i think we can land it.

zekth commented:
So i figured out how to add `/dev/null` and `NUL` reading. But the problem is when permission is not allowed we get a `Deno.ErrorKind.PermissionDenied` the problem is by this way i can't return a `HTTP 403` error code of the fetch local.

Is it revelant to add a feature that without the read permission we may be allowed to read `NUL` and `/dev/null` ? (in this case it's usefull)

ry commented:
I would be very happy to have fetch for local files, but this patch has gone stale. I apologize for not getting it in sooner, @zekth. But in the interest of clearing up old PRs, I'm going to close this.
